car-bridge/src/main.rs

```diff
@@ -29,58 +44,203 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
     let vin = env::<String>(VIN_ENV_NAME);
     let vin = vin.as_deref().unwrap_or(DEFAULT_VIN);
-    let host = env::<String>(BROKER_URL_ENV_NAME);
-    let host = host.as_deref().unwrap_or(DEFAULT_BROKER_URL);
+    let broker_url =
+        env::<Url>(BROKER_URL_ENV_NAME).unwrap_or_else(|| DEFAULT_BROKER_URL.try_into().unwrap());
 
-    info!("Connecting to MQTT broker on '{host}'.");
+    let chariott = GrpcChariott::connect().await?;
 
-    let mut client = AsyncClient::new(
-        CreateOptionsBuilder::new()
-            .mqtt_version(MQTT_VERSION_5)
-            .server_uri(host)
-            .client_id(format!("car-bridge-{vin}"))
-            .finalize(),
-    )?;
+    let mut client = MqttMessaging::connect(broker_url, vin.to_owned()).await?;
+    let mut messages = client.subscribe(format!("c2d/{vin}")).await?;
 
-    let mut message_stream = client.get_stream(MQTT_CLIENT_BUFFER_SIZE);
+    let cancellation_token = ctrl_c_cancellation();
 
-    client.connect(ConnectOptionsBuilder::new().mqtt_version(MQTT_VERSION_5).finalize()).await?;
+    let (response_sender, mut response_receiver) = mpsc::channel(PUBLISH_BUFFER);
 
-    let c2d_topic = format!("c2d/{vin}");
-    info!("Subscribing to topic '{c2d_topic}'.");
-    client.subscribe(c2d_topic.clone(), QOS_2).await?;
+    let publish_handle = {
+        // Detach sending the responses from handling the messages to avoid
+        // backpressure and disconnect the client gracefully.
 
-    // Publish a phony message.
-    client.publish(Message::new(c2d_topic, "Example payload", QOS_2)).await?;
+        let cancellation_token = cancellation_token.child_token();
+        spawn(async move {
+            let client = Arc::new(client);
 
-    let cancellation_token = ctrl_c_cancellation();
+            loop {
```

Comment:
Consider folding this into the loop and `select!` below, which should simplify things and get rid of one `spawn`.

Reply:
Done in c94b8b411264c9100b9b29ef6ef760cccef2688a